homeassistant/components/lifx/light.py

```diff
@@ -376,72 +393,122 @@ async def start_effect(self, entities, service, **kwargs):
         elif service == SERVICE_EFFECT_STOP:
             await self.effects_conductor.stop(bulbs)
 
+    def clear_inflight_discovery(self, inflight: InFlightDiscovery) -> None:
+        """Clear in-flight discovery."""
+
+        if inflight.device.mac_addr in self.discoveries_inflight:
+            self.discoveries_inflight.pop(inflight.device.mac_addr)
+
     @callback
-    def register(self, bulb):
+    def register(self, bulb: Light) -> None:
         """Allow a single in-flight discovery per bulb."""
+        if bulb.mac_addr in self.switch_devices:
+            _LOGGER.debug(
+                "Skipping discovered LIFX Switch at %s (%s)",
+                bulb.ip_addr,
+                bulb.mac_addr,
+            )
+            return
+
         if bulb.mac_addr not in self.discoveries_inflight:
-            self.discoveries_inflight[bulb.mac_addr] = bulb.ip_addr
-            _LOGGER.debug("Discovered %s (%s)", bulb.ip_addr, bulb.mac_addr)
-            self.hass.async_create_task(self.register_bulb(bulb))
+            inflight = InFlightDiscovery(bulb, asyncio.Lock())
+            self.discoveries_inflight[bulb.mac_addr] = inflight
+            _LOGGER.debug(
+                "Discovery response received from %s (%s)",
+                bulb.ip_addr,
+                bulb.mac_addr,
+            )
         else:
-            _LOGGER.warning("Duplicate LIFX discovery response ignored")
+            _LOGGER.debug(
+                "Duplicate discovery response received from %s (%s)",
+                bulb.ip_addr,
+                bulb.mac_addr,
+            )
+
+        self.hass.async_create_task(
+            self._async_handle_discovery(self.discoveries_inflight[bulb.mac_addr])
+        )
 
-    async def register_bulb(self, bulb):
+    async def _async_handle_discovery(self, inflight: InFlightDiscovery) -> None:
         """Handle LIFX bulb registration lifecycle."""
-        if bulb.mac_addr in self.entities:
-            entity = self.entities[bulb.mac_addr]
-            entity.registered = True
-            _LOGGER.debug("Reconnected to %s", entity.who)
-            await entity.update_hass()
-        else:
-            _LOGGER.debug("Connecting to %s (%s)", bulb.ip_addr, bulb.mac_addr)
 
-            # Read initial state
+        # only allow a single discovery process per discovered device
+        async with inflight.lock:
+
```

Comment:
```suggestion
            if entity := self.entities.get(inflight.device.mac_addr):
                # Multiple discoveries in-flight at the same
                # time and this one lost the race and the first
                # one already created the entity
                _LOGGER.debug("Already created %s", entity.who)
                return
```

Reply:
```suggestion
            if inflight.device.mac_addr not in self.discoveries_inflight:
                # Multiple discoveries in-flight at the same
                # time and this one lost the race and the first
                # one already created the entity
                _LOGGER.debug("Already created %s", inflight.device.mac_addr)
                return

```
That might be better though

Reply:
Since one of the reported problems is concurrent discoveries, Here's the exact problem I'm concerned about here:

Discovery 1 happens for bulb A, it wins and gets the lock via `async with inflight.lock:`
Discovery 2 happens for bulb A again, it loses the race and waits for `async with inflight.lock:`
Discovery 3 happens for bulb A again, it loses the race and waits for `async with inflight.lock:`
Discovery 4 happens for bulb A again, it loses the race and waits for `async with inflight.lock:`
Discovery 1 proceeds and creates the entity
Discovery 2 proceeds and the entity is already created so it hits https://github.com/home-assistant/core/pull/72213/files#diff-71a06a9aebb8a8734257fa1b6545e754d5c23ee87d5c05ec92133498c8afea3cR463
Discovery 3 proceeds and the entity is already created so it hits https://github.com/home-assistant/core/pull/72213/files#diff-71a06a9aebb8a8734257fa1b6545e754d5c23ee87d5c05ec92133498c8afea3cR463
Discovery 4 proceeds and the entity is already created so it hits https://github.com/home-assistant/core/pull/72213/files#diff-71a06a9aebb8a8734257fa1b6545e754d5c23ee87d5c05ec92133498c8afea3cR463

We could end up polling the bulb 4 (or more) times in quick succession which could overwhelm the device.

Reply:
It's not so much that concurrent discoveries are a problem, it was an issue with dropping the entity when discovery failed. The next round attempted to recreate the entity, which threw a "doesn't generate unique entity IDs" error. 

This refactor eliminates that error by not dropping the entity from `LIFXManager`'s `self.entities` array. It just marks the entity unavailable. If the user manually deletes the entity, the device is deleted and discovery works just fine if the bulb reappears.

Reply:
So I think add the guard in https://github.com/home-assistant/core/pull/72213#discussion_r894233205 and to return only if entity is already available, and this PR should be good go.